### DISCUSSION TITLE
How do I pass test generated values to AfterEachCallback

### DISCUSSION THREAD

**Discussion by pratyushdeshpande** [junit-team/junit-framework#3613] [Q&A] [answered]
Hello,

I want to pass the values which was generated during the test phase to the AfterEachCallback extension where I need to then cleanup that value. Since my Tests runs parallel, I also want to be thread safe, as it should not clean up value for test which are not completed. Also, I did not find a way to pass the value to AfterEachCallback extension, can someone help ?

**Comment by marcphilipp** [comment, marked as answer]
You can use a `ParameterResolver` to inject an object into the test method that can create the object you need to clean up or register it. The extension implementing `ParameterResolver` and `AfterEachCallback` can then use the `ExtensionContext.Store` to store and retrieve it. If cleanup is all you need, you might not even need to implement `AfterEachCallback` but can wrap your object in `CloseableResource` before putting it into `ExtensionContext.Store` which will cause it to be cleaned up automatically.

**Reply by pratyushdeshpande** [reply]
Thank you for your reply, Can it be done without modifying the already existing Test methods ?

**Reply by marcphilipp** [reply]
How would the extension know which objects it needs to clean up?

**Reply by pratyushdeshpande** [reply]
I was thinking passing the object from beforeEachCallback to the Tests, update the object in the Test and then cleanup in afterEachCallback, but again I could not find how to pass the object from beforeEachCallback to Test itself

**Reply by pratyushdeshpande** [reply]
Also, How would the ExtensionContext.getStore work with different tests running in parallel. I have done something like this: 
in resolveParamter method : 
`extensionContext.getStore(ExtensionContext.Namespace.GLOBAL).getOrComputeIfAbsent(extensionContext.getUniqueId(), (id) -> objectToBeModifiedInTest);`

And in afterEach method : 
`(ObjectToBeModifiedInTest) extensionContext.getStore(ExtensionContext.Namespace.GLOBAL).remove(extensionContext.getUniqueId());`

 Will the above remove the Object modified for that specified test or can it occur that it remove the Object created in other test that was running in parallel ?

**Reply by marcphilipp** [reply]
Every `ExtensionContext` has its own `Store`. Using that takes care of concurrency for you. Please see `TemporaryDirectory` for a complete example that does something very similar.

**Reply by pratyushdeshpande** [reply]
Thank you so much your suggestion, I was able to resolve the issue
